### PR TITLE
add single test for building libumf.so to gha

### PR DESCRIPTION
**Pull request by bratpiorka**

### CODE REVIEW

File: .github/workflows/basic.yml [hunk 1]
```diff
@@ -12,13 +12,18 @@ jobs:
         build_type: [Debug, Release]
         compiler: [{c: gcc, cxx: g++}]
         pool_tracking: ['-DUMF_ENABLE_POOL_TRACKING=ON', '-DUMF_ENABLE_POOL_TRACKING=OFF']
+        shared_library: ['-DUMF_BUILD_SHARED_LIBRARY=OFF']
```
Comment (ldorau) [inline .github/workflows/basic.yml:15]: Could you simplify those options like in https://github.com/oneapi-src/unified-memory-framework/pull/87/files ?

**Review by ldorau — CHANGES_REQUESTED** [review]

**Comment by ldorau** [comment]
It will build when you rebase it on top of https://github.com/oneapi-src/unified-memory-framework/pull/86

File: .github/workflows/basic.yml [hunk 2]
```diff
@@ -12,13 +12,18 @@ jobs:
...
           - os: 'ubuntu-20.04'
             build_type: Release
             compiler: {c: gcc-7, cxx: g++-7}
           - os: 'ubuntu-22.04'
             build_type: Release
-            compiler: {c: clang, cxx: clang++}
+            compiler: {c: clang, cxx: clang++}            
+          - os: 'ubuntu-22.04'
+            build_type: Release
+            compiler: {c: gcc, cxx: g++}
```
Comment (PatKamin) [inline .github/workflows/basic.yml:25]: `compiler` line is redundant, I think. The main matrix doesn't include  other compilers than gcc.

Comment (ldorau) [reply .github/workflows/basic.yml:25]: It is not redundant. It is needed, because it is used here:
```
          -DCMAKE_C_COMPILER=${{matrix.compiler.c}}
          -DCMAKE_CXX_COMPILER=${{matrix.compiler.cxx}}
```

**Review by ldorau — CHANGES_REQUESTED** [review]
Please rebase on the top of the current main branch.

File: .github/workflows/basic.yml (hunk 2 above)
Comment (bratpiorka) [reply .github/workflows/basic.yml:25]: I'm not sure why, but this workflow doesn't work if I omit setting the compiler even if I use the "default" values

File: .github/workflows/basic.yml (hunk 1 above)
Comment (bratpiorka) [reply .github/workflows/basic.yml]: done

File: .github/workflows/basic.yml (hunk 2 above)
Comment (PatKamin) [reply .github/workflows/basic.yml:25]: I see now. Because you added a 'global' `shared_library` variable for multiple jobs and you change this value to a value not present in the matrix, you have to define all variables here as this job's configuration can't be added to existing jobs. See [here](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations )

File: .github/workflows/basic.yml [hunk 3]
```diff
@@ -12,13 +12,18 @@ jobs:
         build_type: [Debug, Release]
         compiler: [{c: gcc, cxx: g++}]
         pool_tracking: ['ON', 'OFF']
+        shared_library: ['OFF']
```
Comment (PatKamin) [inline .github/workflows/basic.yml:15]: Not for this PR: Now when we have this short version of flags variables, there is an information lost in GHA jobs listing. Example title of a job: 
[Build - Ubuntu (ubuntu-22.04, Release, gcc, g++, OFF, OFF)](https://github.com/oneapi-src/unified-memory-framework/actions/runs/7286710419/job/19855967077?pr=83#logs)
Who will know what does each of an `OFF` mean? We should print job matrix configuration somewhere during a build to distinguish failing jobs.

Comment (bratpiorka) [reply .github/workflows/basic.yml:15]: We could always check the "configure build" section in logs where we see full cmake command. But still printing the config matrix is a good idea - please note that we have a task for this UMFW-173

**Review by PatKamin — APPROVED** [review]

**Comment by bratpiorka** [comment]
> Please rebase on the top of the current main branch.
@ldorau done

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -50,6 +50,11 @@ function(add_umf_test)
 
     set_tests_properties(${TEST_NAME} PROPERTIES LABELS "umf")
 
+    if(WIN32)
+        set_property(TEST ${TEST_NAME} PROPERTY ENVIRONMENT_MODIFICATION
```
Comment (igchor) [inline test/CMakeLists.txt:54]: Why is this needed? link_libraries should be enough

**Review by ldorau — CHANGES_REQUESTED** [review]
During testing this PR locally on my WSL I got the following errors:
```sh
[build]$ cmake .. -DUMF_BUILD_SHARED_LIBRARY=ON -DCMAKE_BUILD_TYPE=Debug ; nmake ; ctest --output-on-failure
[...]
Test project C:/Users/ldorau/work/unified-memory-framework/build
    Start 1: umf-base
1/4 Test #1: umf-base .........................   Passed    0.32 sec
    Start 2: umf-memoryPool
2/4 Test #2: umf-memoryPool ...................Exit code 0xc0000135
***Exception:   0.34 sec

    Start 3: umf-memoryProvider
3/4 Test #3: umf-memoryProvider ...............Exit code 0xc0000135
***Exception:   0.37 sec

    Start 4: umf-memory_pool_internal
4/4 Test #4: umf-memory_pool_internal .........Exit code 0xc0000135
***Exception:   0.37 sec


25% tests passed, 3 tests failed out of 4

Label Time Summary:
umf    =   1.40 sec*proc (4 tests)

Total Test time (real) =   1.45 sec

The following tests FAILED:
          2 - umf-memoryPool (Exit code 0xc0000135
)
          3 - umf-memoryProvider (Exit code 0xc0000135
)
          4 - umf-memory_pool_internal (Exit code 0xc0000135
)
Errors while running CTest
```

Probably the same like in the build: https://github.com/oneapi-src/unified-memory-framework/actions/runs/7291816508/job/19871516532?pr=83

**Review by ldorau — APPROVED** [review]

File: test/CMakeLists.txt (hunk 4 above)
Comment (bratpiorka) [reply test/CMakeLists.txt:57]: On Win, link_libraries is used to link to .lib with export declarations, but we have to also provide a path to .dll file which is located in a different folder (e.g. build/lib/Release/umf.lib vs build/bin/Release/umf.dll)

File: include/umf/base.h [hunk 5]
```diff
@@ -17,6 +17,12 @@
 extern "C" {
 #endif
 
+#ifdef _WIN32
+#define UMF_EXPORT __declspec(dllexport)
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply include/umf/base.h:21]: I added an additional #ifdef UMF_SHARED_LIBRARY to define this only when needed

**Review by ldorau — APPROVED** [review]

**Review by igchor — APPROVED** [review]
